src/Http/Environment.php

```diff
@@ -0,0 +1,981 @@
+<?php
+
+namespace Kirby\Http;
+
+use Kirby\Cms\Helpers;
+use Kirby\Exception\InvalidArgumentException;
+use Kirby\Filesystem\F;
+use Kirby\Toolkit\A;
+use Kirby\Toolkit\Str;
+
+/**
+ * The environment object takes care of
+ * secure host and base URL detection, as
+ * well as loading the dedicated
+ * environment options.
+ * @since 3.7.0
+ *
+ * @package   Kirby Http
+ * @author    Bastian Allgeier <bastian@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://opensource.org/licenses/MIT
+ */
+class Environment
+{
+    /**
+     * Whether the request is being served by the CLI
+     *
+     * @var bool
+     */
+    protected $cli;
+
+    /**
+     * Current host name
+     *
+     * @var string
+     */
+    protected $host;
+
+    /**
+     * Whether the HTTPS protocol is used
+     *
+     * @var bool
+     */
+    protected $https;
+
+    /**
+     * Sanitized `$_SERVER` data
+     *
+     * @var array
+     */
+    protected $info;
+
+    /**
+     * Current server's IP address
+     *
+     * @var string
+     */
+    protected $ip;
+
+    /**
+     * Whether the site is behind a reverse proxy;
+     * `null` if not known (fixed allowed URL setup)
+     *
+     * @var bool|null
+     */
+    protected $isBehindProxy;
+
+    /**
+     * URI path to the base
+     *
+     * @var string
+     */
+    protected $path;
+
+    /**
+     * Port number in the site URL
+     *
+     * @var int|null
+     */
+    protected $port;
+
+    /**
+     * Intermediary value of the port
+     * extracted from the host name
+     *
+     * @var int|null
+     */
+    protected $portInHost;
+
+    /**
+     * Array with path and query
+     * from the `REQUEST_URI`
+     *
+     * @var array
+     */
+    protected $requestUri;
+
+    /**
+     * Path to the php script within the
+     * document root without the
+     * filename of the script
+     *
+     * @var string
+     */
+    protected $scriptPath;
+
+    /**
+     * Full base URL
+     *
+     * @var string
+     */
+    protected $url;
+
+    /**
+     * Class constructor
+     *
+     * @param array|null $options
+     * @param array|null $info Optional override for `$_SERVER`
+     */
+    public function __construct(?array $options = null, ?array $info = null)
+    {
+        $this->detect($options, $info);
```

Comment:
This line (direct detection from the `Environment` constructor) causes a bug in the error handling.

**How to reproduce:**

When you define the `url` option e.g. as:

```php
<?php

return [
    'debug' => true,
    'url'   => ['https://getkirby.com', 'https://trykirby.com']
];
```

and visit from a different domain, the `Environment` class throws the "The environment is not allowed" exception. But the user only sees the fatal view, even if `debug` is enabled.

**Reason for the bug:**

This is because when `App::optionsFromEnvironment()` creates the `Environment` instance, the detection runs before the instance is assigned to `$app->environment`. So by the time the exception is finally thrown, `$app->environment` is still `null`. This means that `AppErrors::handleErrors()` cannot access the CLI status from the environment, which throws another PHP error. So the second fatal error triggers the fatal view.

**Possible fix:**

I think we can only fix this by removing the constructor and only calling `detect()` manually from `App::optionsFromEnvironment()`. On the other hand this has the downside that the `Environment` class can only be used in this way. Calling `detect()` can easily be forgotten.

Reply:
Couldn't we fix it with a similar approach like in the URI class? If the environment is already detected, we use that instance. If it's still null, we use a new "detached" environment instance.

Reply:
I think an `Environment` singleton could already solve this. The `Environment::instance()` method could create the instance, set it to its static singleton property and then trigger the detection.

Reply:
I solved it a bit differently. Rethinking it, another cached singleton would probably lead to more confusion down the road.